### PR TITLE
Custom ignored functions in stacktrace

### PR DESCRIPTION
**Pull request by Avgor46**

### CODE REVIEW

File: src/bin/casr-python.rs [hunk 1]
```diff
@@ -104,6 +105,13 @@ fn main() -> Result<()> {
                 .value_name("FILE")
                 .help("Stdin file for program"),
         )
+        .arg(
+            Arg::new("ignore")
```
Comment (SweetVishnya) [inline src/bin/casr-python.rs:109]: Also, pass ignore parameter in `call_casr_san` function when calling casr-san.

File: src/error.rs [hunk 2]
```diff
@@ -45,4 +45,10 @@ impl From<goblin::error::Error> for Error {
     }
 }
 
+impl From<anyhow::Error> for Error {
+    fn from(err: anyhow::Error) -> Error {
+        Error::Casr(err.chain().rev().map(|s| s.to_string() + ". ").collect())
```
Comment (SweetVishnya) [inline src/error.rs:50]: Why do we need this? Why do we change anyhow error to Casr?

File: src/stacktrace_constants.rs [hunk 3]
```diff
@@ -221,3 +223,10 @@ pub const STACK_FRAME_FILEPATH_IGNORE_REGEXES: &[&str] = &[
     r".*libubsan\.so",
     r".*asan_with_fuzzer\.so",
 ];
+
+lazy_static::lazy_static! {
+    pub static ref STACK_FRAME_FUNCION_IGNORE_REGEXES_MUTABLE: RwLock<Vec<String>> = RwLock::new(
```
Comment (SweetVishnya) [inline src/stacktrace_constants.rs:228]: docs for variables

File: tests/casr_tests/test_casr_ignore_frames/ign1.lst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+FUNCTIONS
```
Comment (SweetVishnya) [inline tests/casr_tests/test_casr_ignore_frames/ign1.lst:1]: We need documentation for ign.lst format. Update usage.md.

File: src/util.rs [hunk 5]
```diff
@@ -128,3 +132,35 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
     }
     Ok(())
 }
+
+pub fn change_ignored_frames(path: &Path) -> error::Result<()> {
```
Comment (SweetVishnya) [inline src/util.rs:136]: Maybe we should move it to stacktrace_constants.rs?

File: src/util.rs [hunk 6]
```diff
@@ -128,3 +132,35 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
...
+    let file = std::fs::File::open(path)
+        .with_context(|| format!("Cannot open file: {}", path.display()))?;
+    let mut reader = BufReader::new(file)
+        .lines()
+        .map(|x| x.unwrap())
+        .collect::<Vec<String>>();
+    if reader.is_empty() || !reader[0].contains("FUNCTIONS") && !reader[0].contains("FILES") {
+        return Err(error::Error::Casr(format!(
+            "File {} is empty or does not contain \
+                    FUNCTION or FILES on the first line",
```
Comment (SweetVishnya) [inline src/util.rs:146]: FUNCTIONS

File: src/util.rs [hunk 7]
```diff
@@ -128,3 +132,35 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
...
+            path.display()
+        )));
+    }
+    let (funcs, paths) = if reader[0].contains("FUNCTIONS") {
+        if let Some(bound) = reader.iter().position(|x| x.contains("FILES")) {
+            let files = reader.split_off(bound);
+            (reader, files)
+        } else {
+            (reader, vec![r"^[^.]$".to_string()])
```
Comment (SweetVishnya) [inline src/util.rs:155]: We'd better (reader, None)

File: src/util.rs [hunk 8]
```diff
@@ -128,3 +132,35 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
...
+        }
+    } else if let Some(bound) = reader.iter().position(|x| x.contains("FUNCTIONS")) {
+        let funcs = reader.split_off(bound);
+        (funcs, reader)
+    } else {
+        (vec![r"^[^.]$".to_string()], reader)
```
Comment (SweetVishnya) [inline src/util.rs:161]: (None, reader)

File: src/util.rs [hunk 9]
```diff
@@ -128,3 +132,35 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
...
+        return Err(error::Error::Casr(format!(
+            "File {} is empty or does not contain \
+                    FUNCTION or FILES on the first line",
+            path.display()
+        )));
+    }
+    let (funcs, paths) = if reader[0].contains("FUNCTIONS") {
+        if let Some(bound) = reader.iter().position(|x| x.contains("FILES")) {
+            let files = reader.split_off(bound);
+            (reader, files)
```
Comment (SweetVishnya) [inline src/util.rs:153]: We shouldn't replace filters with the user specified once. Instead we should append user filters to default filters.

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: src/bin/casr-python.rs (hunk 1 above)
Comment (Avgor46) [reply src/bin/casr-python.rs:112]: done

File: src/error.rs (hunk 2 above)
Comment (Avgor46) [reply src/error.rs:50]: For convenient use `?`:
image

File: src/stacktrace_constants.rs (hunk 3 above)
Comment (Avgor46) [reply src/stacktrace_constants.rs:229]: done

File: tests/casr_tests/test_casr_ignore_frames/ign1.lst (hunk 4 above)
Comment (Avgor46) [reply tests/casr_tests/test_casr_ignore_frames/ign1.lst:1]: done

File: src/util.rs (hunk 6 above)
Comment (Avgor46) [reply src/util.rs]: done

File: src/util.rs (hunk 7 above)
Comment (Avgor46) [reply src/util.rs]: (reader, vec![])

File: src/util.rs (hunk 8 above)
Comment (Avgor46) [reply src/util.rs]: (vec![], reader)

File: src/util.rs (hunk 9 above)
Comment (Avgor46) [reply src/util.rs:153]: done

File: src/util.rs (hunk 5 above)
Comment (Avgor46) [reply src/util.rs:136]: stacktrace_constants.rs now contains only constants and variables. Maybe we shouldn't move the function?

File: src/util.rs [hunk 10]
```diff
@@ -128,3 +132,41 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
     }
     Ok(())
 }
+
+pub fn change_ignored_frames(path: &Path) -> error::Result<()> {
```
Comment (anfedotoff) [inline src/util.rs:136]: We need to rename this. Maybe "add_custom_ignored_frames"? Return  just Result<()>.

File: src/error.rs [hunk 11]
```diff
@@ -45,4 +45,10 @@ impl From<goblin::error::Error> for Error {
     }
 }
 
+impl From<anyhow::Error> for Error {
```
Comment (anfedotoff) [inline src/error.rs:48]: We don't need this we could return anyhow::Error

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 12]
```diff
@@ -7,3 +7,14 @@ pub mod python;
 pub mod report;
 pub mod stacktrace_constants;
 pub mod util;
+
+// This macro merges all [&str] slices into single Vec<String>.
+#[macro_export]
+macro_rules! concatall {
```
Comment (anfedotoff) [inline src/lib.rs:13]: Maybe we put this macro to stacktrace_constants/utill? And think about better name?

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: src/stacktrace_constants.rs [hunk 13]
```diff
@@ -222,3 +243,12 @@ pub const STACK_FRAME_FILEPATH_IGNORE_REGEXES: &[&str] = &[
     r".*libubsan\.so",
     r".*asan_with_fuzzer\.so",
 ];
+
+lazy_static::lazy_static! {
+    // Mutable variable for custom ignored function names
```
Comment (anfedotoff) [inline src/stacktrace_constants.rs:248]: We need a new comment here:
Regular expressions for functions to be ignored.

File: src/stacktrace_constants.rs [hunk 14]
```diff
@@ -222,3 +243,12 @@ pub const STACK_FRAME_FILEPATH_IGNORE_REGEXES: &[&str] = &[
...
+    pub static ref STACK_FRAME_FUNCTION_IGNORE_REGEXES: RwLock<Vec<String>> = RwLock::new(
+        Vec::new());
+    // Mutable variable for custom ignored file paths
```
Comment (anfedotoff) [inline src/stacktrace_constants.rs:251]: Regular expressions for file paths to be ignored.

**Review by anfedotoff — APPROVED** [review]

File: src/stacktrace_constants.rs (hunk 14 above)
Comment (Avgor46) [reply src/stacktrace_constants.rs]: done

**Review by anfedotoff — APPROVED** [review]

File: src/lib.rs (hunk 12 above)
Comment (Avgor46) [reply src/lib.rs]: done

**Pull request merged**